bundles/org.openhab.binding.upnpcontrol/src/main/java/org/openhab/binding/upnpcontrol/internal/handler/UpnpHandler.java

```diff
@@ -12,55 +12,281 @@
 import org.openhab.binding.upnpcontrol.internal.config.UpnpControlConfiguration;
+import org.openhab.binding.upnpcontrol.internal.queue.UpnpPlaylistsListener;
+import org.openhab.binding.upnpcontrol.internal.util.UpnpControlUtil;
+import org.openhab.core.common.ThreadPoolManager;
 import org.openhab.core.io.transport.upnp.UpnpIOParticipant;
 import org.openhab.core.io.transport.upnp.UpnpIOService;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
 import org.openhab.core.thing.Thing;
 import org.openhab.core.thing.ThingStatus;
 import org.openhab.core.thing.ThingStatusDetail;
 import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.binding.builder.ThingBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.CommandDescription;
+import org.openhab.core.types.CommandDescriptionBuilder;
+import org.openhab.core.types.CommandOption;
+import org.openhab.core.types.StateDescription;
+import org.openhab.core.types.StateDescriptionFragmentBuilder;
+import org.openhab.core.types.StateOption;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * The {@link UpnpHandler} is the base class for {@link UpnpRendererHandler} and {@link UpnpServerHandler}.
+ * The {@link UpnpHandler} is the base class for {@link UpnpRendererHandler} and {@link UpnpServerHandler}. The base
+ * class implements UPnPConnectionManager service actions.
  *
  * @author Mark Herwege - Initial contribution
  * @author Karel Goderis - Based on UPnP logic in Sonos binding
  */
 @NonNullByDefault
-public abstract class UpnpHandler extends BaseThingHandler implements UpnpIOParticipant {
+public abstract class UpnpHandler extends BaseThingHandler implements UpnpIOParticipant, UpnpPlaylistsListener {
 
     private final Logger logger = LoggerFactory.getLogger(UpnpHandler.class);
 
-    protected UpnpIOService service;
-    protected volatile String transportState = "";
-    protected volatile int connectionId;
-    protected volatile int avTransportId;
-    protected volatile int rcsId;
-    protected @NonNullByDefault({}) UpnpControlConfiguration config;
+    // UPnP constants
+    static final String CONNECTION_MANAGER = "ConnectionManager";
+    static final String CONNECTION_ID = "ConnectionID";
+    static final String AV_TRANSPORT_ID = "AVTransportID";
+    static final String RCS_ID = "RcsID";
+    static final Pattern PROTOCOL_PATTERN = Pattern.compile("(?:.*):(?:.*):(.*):(?:.*)");
 
-    public UpnpHandler(Thing thing, UpnpIOService upnpIOService) {
+    protected UpnpIOService upnpIOService;
+
+    protected volatile @Nullable RemoteDevice device;
+
+    // The handlers can potentially create an important number of tasks, therefore put them in a separate thread pool
+    protected ScheduledExecutorService upnpScheduler = ThreadPoolManager.getScheduledPool("binding-upnpcontrol");
```

Comment:
The number of tasks is not that important, but the duration of those tasks. Do you schedule any long running tasks?

Reply:
Many of them wait for a response from the device, so these could take some time. Also, it made it a lot easier to debug, as it is possible to more easily identify the threads started by the handlers.

Reply:
These are acutally no valid arguments to create an own thread pool. @cpmeister WDYT?

Reply:
He uses `invokeActionLock` to run one of the tasks essentially single threaded, which although would be more efficient to use a single-threaded executor for that section I think a separate threadpool should be used regardless because you don't want to eat up threads on the shared threadpool for such a task.

Reply:
OK